File: 2024/day2/day2.py
import sys

def is_safe(report: list[int]) -> bool:
    if all(1 <= y - x <= 3 for x, y in zip(report, report[1:])):
        return True
    if all(1 <= x - y <= 3 for x, y in zip(report, report[1:])):
        return True
    return False

def is_tolerably_safe(report: list[int]) -> bool:
    return any(is_safe(report[:i] + report[i+1:]) for i in range(len(report)))

reports = [list(map(int, line.split())) for line in sys.stdin]
print(f'Part 1: {sum(is_safe(report) for report in reports)}')
print(f'Part 2: {sum(is_tolerably_safe(report) for report in reports)}')

File: 2024/day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func isSafe(report []int) bool {
	ascending, descending := true, true

	for i := 1; i < len(report); i++ {
		diff := report[i] - report[i-1]
		if diff < 1 || diff > 3 {
			ascending = false
		}
		if -diff < 1 || -diff > 3 {
			descending = false
		}
	}

	return ascending || descending
}

func isTolerablySafe(report []int) bool {
	for i := 0; i < len(report); i++ {
		// Create a copy of the report excluding the i-th element
		modified := append([]int{}, report[:i]...)
		modified = append(modified, report[i+1:]...)
		if isSafe(modified) {
			return true
		}
	}
	return false
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var reports [][]int

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		report := make([]int, len(parts))
		for i, part := range parts {
			num, _ := strconv.Atoi(part)
			report[i] = num
		}
		reports = append(reports, report)
	}

	part1, part2 := 0, 0
	for _, report := range reports {
		if isSafe(report) {
			part1++
		}
		if isTolerablySafe(report) {
			part2++
		}
	}

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}
